src/wasm/wasm-binary.cpp

```diff
@@ -2780,6 +2826,30 @@ void WasmBinaryBuilder::readNames(size_t payloadLen) {
                     << std::to_string(index) << std::endl;
         }
       }
+    } else if (nameType == BinaryConsts::UserSections::Subsection::NameField) {
+      auto numTypes = getU32LEB();
+      for (size_t i = 0; i < numTypes; i++) {
+        auto index = getU32LEB();
+        if (index >= types.size() || !types[index].isStruct()) {
+          std::cerr << "warning: invalid type index in name field section\n";
+        } else {
+          auto& fields = types[index].getStruct().fields;
+          auto numFields = getU32LEB();
+          for (size_t i = 0; i < numFields; i++) {
+            auto index = getU32LEB();
+            if (index >= fields.size()) {
+              std::cerr
+                << "warning: invalid field index in name field section\n";
+            } else {
+              // getStruct() intentionally returns a constant result. However,
+              // here we just adjust the name (which is not part of the hashing
+              // etc.) which is ok to do.
+              auto* field = const_cast<Field*>(&fields[index]);
```

Comment:
A few issues come to mind: First, if someone uses Binaryen as a library and parses multiple modules in parallel, there will be a data race here. Second, any use of Binaryen that parses multiple modules (not necessarily in parallel) will only end up with the names from one of the modules. Those are perhaps esoteric problems, but what if we kept the field names in a side table owned by the `Module` instead of making them part of the globally interned types? That would solve both of these problems and would also scale to preserving the names of HeapTypes as well, which I assume we will want to do as we parse and print wasts with complex types.

Reply:
Yes, I think you're right. We keep local names in a map on the side in Functions, so this would be consistent with that. I'll look into a refactoring for it.